### PR TITLE
LSRA: remove last uses only at use point

### LINKED ISSUE #17389: AVX2_ro fails with jitstress=2
Forked from #17027 / #17330.

x64 checked build. AVX2_ro fails to validate with jitstress=2. The `BlendVariable<Byte>` are the ones failing.
```
Avx2.BlendVariable<Byte>(Vector256<Byte>, Vector256<Byte>, Vector256<Byte>): RunBasicScenario_Load failed:
   firstOp: (63, 251, 58, 216, 203, 64, 178, 128, 158, 102, 241, 93, 117, 12, 119, 40, 204, 56, 100, 37, 64, 108, 251, 16, 7, 29, 119, 127, 144, 21, 114, 30)
  secondOp: (151, 55, 219, 192, 119, 228, 54, 168, 120, 32, 66, 32, 41, 148, 50, 0, 40, 175, 220, 151, 239, 227, 58, 209, 0, 41, 208, 173, 200, 50, 238, 136)
   thirdOp: (128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1, 128, 1)
    result: (151, 251, 219, 216, 119, 64, 54, 128, 120, 102, 66, 93, 41, 12, 50, 40, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
... etc ... 
```

cc @dotnet/jit-contrib @RussKeldorph @fiigii

### PR DESCRIPTION
**Pull request by CarolEidt**
LSRA maintains liveness within a block to determine what's live across a call. It uses the last use bits on lclVar nodes to remove them from the set. However, this should be done at the point of use rather than at the point where the lclVar is encountered in the execution stream.

Fix #17389

### CODE REVIEW

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL
cc @tannergooding @fiigii @4creators

File: src/jit/lsrabuild.cpp [hunk 1]
```diff
@@ -1595,9 +1580,31 @@ void LinearScan::buildRefPositionsForNode(GenTree* tree, BasicBlock* block, Lsra
 
         assert((candidates & allRegs(srcInterval->registerType)) != 0);
 
-        // For non-localVar uses we record nothing, as nothing needs to be written back to the tree.
-        GenTree* const refPosNode = srcInterval->isLocalVar ? useNode : nullptr;
-        RefPosition*   pos        = newRefPosition(srcInterval, currentLoc, RefTypeUse, refPosNode, candidates, 0);
+        GenTree* refPosNode;
+        if (srcInterval->isLocalVar)
+        {
+            // We have only approximate last-use information at this point.  This is because the
```
Comment (sandreenko) [inline src/jit/lsrabuild.cpp]: conditions before were:
` if (tree->OperIs(GT_LCL_VAR, GT_LCL_FLD)) and if (isCandidateVar(varDsc))`
now we have:
`if (srcInterval->isLocalVar) `
are they equal?

Comment (CarolEidt) [reply src/jit/lsrabuild.cpp]: Yes, they are equivalent- at this point we already have `srcInterval`, where before (when looking at the lclVar itself, as opposed to its use), we don't yet have the `Interval` available.

**Comment by CarolEidt** [comment]
> It looks like a big change, do we expect diffs from it? As I understand it makes live intervals longer and can causes more spills.

This will not change the span of the intervals. It will only cause intervals to be reported live if they were incorrectly reported as not live previously. What I didn't explain earlier, but should have, is that for the most part the `currentLiveVars` are used only for preferencing, but for the upper vector saves & restores it uses them to determine which large vectors are live across a call.

It's pretty unusual to have a lclVar in the execution stream before a call, when it's not used/consumed until after the call.

I am now running x64 diffs just to check, but I would be very surprised if there were any.

**Comment by CarolEidt** [comment]
> I am now running x64 diffs just to check, but I would be very surprised if there were any.

I have to eat my words. Naturally (as I should have realized) changing the liveness used for preferencing will have an effect on register allocation. While the net diffs were very small (0.00% as reported by jit-diff), and were an improvement for the frameworks, there were some surprising regressions for the tests. Tracking this down, I realized that, although tracking the uses at their actual point of use was more correct, it causes us to miss contained uses. Generally, register-candidate lclVars aren't contained, but RMW indirections are an exception. The second commit includes a fix to ensure that we observe a last use in such cases. Note that this isn't a correctness issue (only the large vectors require correct liveness in the build phase - and that should really be fixed with #17481) and will never be contained in an addressing mode, and the impact is small overall, but it takes the overall diffs (frameworks + tests):
- From 579 bytes regression to 49 bytes improvement on x86
- From 250 bytes regression to 84 bytes improvement on x64

There are still some small regressions, due to using an extra callee-save register in methods with multiple epilogs (the size of the save/restore code exceeds the savings from avoiding a spill).

**Comment by CarolEidt** [comment]
@dotnet-bot test Windows_NT x64 Checked jitstress2

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib ping

**Comment by CarolEidt** [comment]
> The first commit is correctness fix;
> The second commit is CQ;

Yes, that's correct.

> I do not see any reads of buildRefPositionsForNode between the old and the new positions, so looks like the point where we do this is not important, is it? I think maybe this PR change not when we add smth to this set, but also change what values we add there?

I don't understand the question. Previously, we were removing a last-use lclVar from `currentLiveVars` at the time we were handling the lclVar node in `buildRefPositionsForNode`. Now were are doing that at the time we are handling the node that *uses* that lclVar, which is the actual point of use. We had a case where a lclVar node appeared in the execution stream before a nested call, even though it wasn't used until after that call.

The comment you refer to *is* addressing the correctness question. The out-of-order use issue only arises, AFAIK, in the context of a single expression, e.g. `ADD(V01,V01)` - where the `V01` node marked last use may be the one that occurs earlier in the instruction stream. That's why the comment says:

"the cases where these out-of-order uses occur should not overlap a kill (they are only known to occur within a single expression)."

I firmly believe that we need to revamp the handling of the large vector saves & restores, but I do not believe it is a correctness issue.

**Pull request merged**